cmd/contour/shutdownmanager.go

```diff
@@ -0,0 +1,132 @@
+		}
+		time.Sleep(s.checkInterval)
+	}
+}
+
+// shutdownEnvoy sends a POST request to /healthcheck/fail to tell Envoy to start draining connections
+func shutdownEnvoy(url string) error {
+	resp, err := http.Post(url, "", nil)
+	if err != nil {
+		return fmt.Errorf("creating POST request for URL %q failed: %s", url, err)
+	}
+
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("POST request for URL %q returned HTTP status %s", url, resp.Status)
+	}
+	return nil
+}
+
+// getOpenConnections parses a http request to a prometheus endpoint returning the sum of values found
+func getOpenConnections(url, prometheusStat string, prometheusValues []string) (int, error) {
+	// Make request to Envoy Prometheus endpoint
+	resp, err := http.Get(url)
+	if err != nil {
+		return -1, fmt.Errorf("GET request for URL %q failed: %s", url, err)
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		return -1, fmt.Errorf("GET request for URL %q returned HTTP status %s", url, resp.Status)
+	}
+
+	// Parse Prometheus listener stats for open connections
+	return metrics.ParseOpenConnections(resp.Body, prometheusStat, prometheusValues)
+}
+
+func doShutdownManager(config *shutdownmanagerContext) error {
+	var g workgroup.Group
+
+	g.Add(func(stop <-chan struct{}) error {
+		config.Info("started envoy shutdown manager")
+		defer config.Info("stopped")
+
+		http.HandleFunc("/healthz", config.healthzHandler)
+		http.HandleFunc("/shutdown", config.shutdownHandler)
+		log.Fatal(http.ListenAndServe(fmt.Sprintf(":%d", config.httpServePort), nil))
+
+		return nil
+	})
+
+	return g.Run()
+}
+
+// registerShutdownManager registers the envoy shutdown sub-command and flags
+func registerShutdownManager(cmd *kingpin.CmdClause, log logrus.FieldLogger) (*kingpin.CmdClause, *shutdownmanagerContext) {
+	ctx := &shutdownmanagerContext{
+		FieldLogger: log,
+	}
+	shutdownmgr := cmd.Command("shutdown-manager", "Start envoy shutdown-manager.")
+	shutdownmgr.Flag("check-interval", "Time to poll Envoy for open connections.").Default("5s").DurationVar(&ctx.checkInterval)
+	shutdownmgr.Flag("check-delay", "Time wait before polling Envoy for open connections.").Default("60s").DurationVar(&ctx.checkDelay)
+	shutdownmgr.Flag("min-open-connections", "Min number of open connections when polling Envoy.").Default("0").IntVar(&ctx.minOpenConnections)
+	shutdownmgr.Flag("serve-port", "Port to serve the http server on.").Default("8090").IntVar(&ctx.httpServePort)
+	shutdownmgr.Flag("prometheus-path", "The path to query Envoy's Prometheus HTTP Endpoint.").Default("/stats/prometheus").StringVar(&ctx.prometheusPath)
+	shutdownmgr.Flag("prometheus-stat", "Prometheus stat to query.").Default("envoy_http_downstream_cx_active").StringVar(&ctx.prometheusStat)
+	shutdownmgr.Flag("prometheus-values", "Prometheus values to look for in prometheus-stat.").Default("ingress_http", "ingress_https").StringsVar(&ctx.prometheusValues)
```

Comment:
Perhaps these should not be configurable. Internally, if we use the `ENVOY_HTTP_LISTENER` and `ENVOY_HTTPS_LISTENER` constants, that will let future maintainers know there is a dependency. Since users can't change the listener names that Contour uses, I don't think they can usefully change these settings either.

Reply:
I missed this in my review, but I tend to agree, there's no real reason for these to be configurable by the end user, let's not expose it unless we have to.

Reply:
Yeah, when I added this I went back & forth in my head. I landed on this because I thought maybe users might want to watch a different value than that we are looking at. But that's not a real use-case now, so will remove.